pkg/monitors/http/http.go

```diff
@@ -0,0 +1,254 @@
+	// Ignore certificate error which will be checked after
+	m.conf.SkipVerify = true
+	if m.conf.Regex != "" {
+		// Compile regex
+		m.regex, err = regexp.Compile(m.conf.Regex)
+		if err != nil {
+			m.logger.WithError(err).Error("failed to compile regular expression")
+		}
+	}
+	// Start the metric gathering process here
+	var ctx context.Context
+	ctx, m.cancel = context.WithCancel(context.Background())
+	utils.RunOnInterval(ctx, func() {
+		// get stats for each website
+		for _, site := range m.conf.URLs {
+			logger := m.logger.WithFields(logrus.Fields{"url": site})
+			_, err := url.Parse(site)
+			if err != nil {
+				logger.WithError(err).Error("could not parse url, ignore this url")
+				continue
+			}
+			dps, lastURL, err := m.getHTTPStats(site, logger)
+			if err == nil {
+				parsedURL, _ := url.Parse(lastURL)
+				if parsedURL.Scheme == "https" {
+					tlsDps, err := m.getTLSStats(parsedURL, logger)
+					if err == nil {
+						dps = append(dps, tlsDps...)
+					} else {
+						logger.WithError(err).Error("Failed gathering TLS stats")
+					}
+				}
+			} else {
+				logger.WithError(err).Error("Failed gathering HTTP stats, ignore other stats")
+			}
+			// for _, v := range dps {
+			// 	logger.Warn(v.Metric, v.Value, v.Dimensions)
+			// }
+			m.Output.SendDatapoints(dps...)
+		}
+	}, time.Duration(conf.IntervalSeconds)*time.Second)
+
+	return nil
+}
+
+// Shutdown the monitor
+func (m *Monitor) Shutdown() {
+	// Stop any long-running go routines here
+	if m.cancel != nil {
+		m.cancel()
+	}
+}
+
+func (m *Monitor) getTLSStats(site *url.URL, logger *logrus.Entry) (dps []*datapoint.Datapoint, err error) {
+	// use as an fmt.Stringer
+	host := site.Hostname()
+	port := site.Port()
+	valid := true
+	var secondsLeft float64
+	if port == "" {
+		port = "443"
+	}
+	ipConn, err := net.Dial("tcp", host+":"+port)
+	if err != nil {
+		return
+	}
+	defer ipConn.Close()
+	tlsCfg := &tls.Config{
+		InsecureSkipVerify: m.conf.SkipVerify,
+		ServerName:         host,
+	}
+	if _, err := auth.TLSConfig(tlsCfg, m.conf.CACertPath, m.conf.ClientCertPath, m.conf.ClientKeyPath); err != nil {
+		return nil, err
+	}
+	conn := tls.Client(ipConn, tlsCfg)
+	if err != nil {
+		return
+	}
+	defer conn.Close()
+	err = conn.Handshake()
+	if err != nil {
+		logger.WithError(err).Error("failed during handshake")
+		valid = false
+	}
+	certs := conn.ConnectionState().PeerCertificates
+	for i, cert := range certs {
+		opts := x509.VerifyOptions{
+			Intermediates: x509.NewCertPool(),
+		}
+		if i == 0 {
+			opts.DNSName = host
+			for j, cert := range certs {
+				if j != 0 {
+					opts.Intermediates.AddCert(cert)
+				}
+			}
+			secondsLeft = time.Until(cert.NotAfter).Seconds()
+		}
+		_, err := cert.Verify(opts)
+		if err != nil {
+			logger.WithError(err).Info("failed verify certificate")
+			valid = false
+		}
+	}
+	dimensions := map[string]string{
+		"url":        site.String(),
+		"serverName": host,
+		"isValid":    strconv.FormatBool(valid),
```

Comment:
`isValid` is a metric value, not a dimension.  A dimension should only be uniquely identifying attributes of a timeseries, not an accidental property of it.  You should just have another metric called something like `http.cert_valid`.

Reply:
ok it was my first implementation and it seems to be better like this in my mind.
I tried to explain why in the documentation, using dimensions will ease a generic detector implementation.

I am not sure there is a demarcated border between what info we can push as dimension and what can push as metrics. 

take the aws metric as example we can create a detector which will raise alert on terminated ec2 with something like a count on any ec2 metric and `filter=filter('aws_state', 'terminated')`.
in this case this is the same, the valuable info we need and use to alert is the value of a dimension, not the metric itself.

I understand this is not the common way to do but it does not seem to be an incompatible or problematic way to go.
That being said, I don't have any strong constraint against coming back to my first implementation with the 2 additional metrics if you think I should but:
- I just would like to explain what
- I am curious if I forget anything about this way which could cause problem

Reply:
That `aws_state` value is a _property_ on the `AWSUniqueId` dimension, and not a _dimension_.  See [Metric Metadata](https://docs.signalfx.com/en/latest/metrics-metadata/metrics-metadata.html), especially about a page into that doc and down where it has the table.  The key distinction is that the `aws_state` can change without creating a distinct timeseries for each unique value of the `aws_state`.  Whereas, with the way you have `isValid`, it will make new, non-overlapping timeseries when its value changes.

The agent does have the ability to set properties on dimensions, but I think the more appropriate thing to do here would be to just have separate timeseries.  If it were a property and you wanted to make a detector that fired whenever a cert was found to be invalid, regardless of how much time was remaining until expiry, you could do it, but it would be hacky because you would have to just do a count on the number of timeseries for `http.cert_expiry` with a filter for `isValid == "false"`.  If it is a separate timeseries you can just have a detector for `http.cert_valid` with analytics that excludes `value != 0`.  For your `http.cert_expiry` detector, I wouldn't worry about whether the cert is valid or not there, just have two detectors: one that checks expiry time and another that checks validity.  Especially since expiry is really only one aspect of whether a cert is valid; it could be invalid due to many other factors.

Does this make sense?  Is there any other aspect of this in your use case that I have not understood?

Reply:
it makes sens, thank you for the explanation.
it is little unfortunate we cannot set property from monitor this is exactly what I look for to "log" some contextual / changing information. (i.e. the SSL error for cert_valid metric).

in any case, I just only need to be careful to report those new separate metrics only when it makes sens and this should be easy to create generic detectors (what we do not want is to have regex not matched alert when we do not configure regex for example).

and I prefer to follow your advices so here the changes for code and doc but I need to update tests now.

Reply:
this time I resolve, everything should be good now @keitwb